// simple-is/simple-is-compiled.js
"use strict";

function _typeof(obj) { return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj; }

// simple-is.js
// MIT licensed, see LICENSE file
// Copyright (c) 2013 Olov Lassus <olov.lassus@gmail.com>

var is = (function () {
    "use strict";

    var hasOwnProperty = Object.prototype.hasOwnProperty;
    var toString = Object.prototype.toString;
    var _undefined = void 0;

    return {
        nan: function nan(v) {
            return v !== v;
        },
        boolean: function boolean(v) {
            return typeof v === "boolean";
        },
        number: function number(v) {
            return typeof v === "number";
        },
        string: function string(v) {
            return typeof v === "string";
        },
        fn: function fn(v) {
            return typeof v === "function";
        },
        object: function object(v) {
            return v !== null && (typeof v === "undefined" ? "undefined" : _typeof(v)) === "object";
        },
        primitive: function primitive(v) {
            var t = typeof v === "undefined" ? "undefined" : _typeof(v);
            return v === null || v === _undefined || t === "boolean" || t === "number" || t === "string";
        },
        array: Array.isArray || function (v) {
            return toString.call(v) === "[object Array]";
        },
        finitenumber: function finitenumber(v) {
            return typeof v === "number" && isFinite(v);
        },
        someof: function someof(v, values) {
            return values.indexOf(v) >= 0;
        },
        noneof: function noneof(v, values) {
            return values.indexOf(v) === -1;
        },
        own: function own(obj, prop) {
            return hasOwnProperty.call(obj, prop);
        }
    };
})();

if (typeof module !== "undefined" && typeof module.exports !== "undefined") {
    module.exports = is;
}
