// tsheets/lib/base.js
;(function () {
  'use strict';

  var klass = require('klass');
  var request = require('request');

  var Base = klass(function() {
    this._config = null;
    this._baseUrl = 'https://rest.tsheets.com/api/v1';
  })
    .methods({
      buildUrl: function buildUrl(url, obj) {
        if (!url) {
          throw new Error('`url` argument is missing from `Base.buildUrl` method.');
        }

        if (!obj) {
          throw new Error('`obj` argument is missing from `Base.buildUrl` method.');
        }
        return url + '?'+Object.keys(obj).reduce(function(a,k){a.push(k+'='+encodeURIComponent(obj[k]));return a},[]).join('&');
      },
      request: function request(url, method, opts, fn) {
        (opts = opts || {});
        opts.client_id = opts.client_id || this.getConfig('client_id');
        opts.client_secret = opts.client_secret || this.getConfig('client_secret');
        (method = method || 'GET');
        request({
          uri: url,
          method: method,
          formData: opts
        }, function(err, httpResponse, body) {
          if (err) {
            return fn(err);
          }

          try {
            body = JSON.parse(body);
          } finally {
            return fn(null, body);
          }
        });
      },
      setConfig: function setConfig(config) {
        this._config = config;
      },
      getConfig: function() {
        return this._config;
      }
    });

    module.exports = Base;
}());
